### PR TITLE
Do staking rewiev

### PR DESCRIPTION
**Pull request by SKonstantinS**
All about staking contract 
something like a documentation in readme.md

### CODE REVIEW

File: staking/contracts/Staking.sol [hunk 1]
```diff
@@ -4,87 +4,250 @@ pragma solidity >=0.6.2 <0.8.0;
...
+        uint blockTime = roundTimestamp(block.timestamp);
+        verification(account, id, newAmount, newSlope, newCliff, blockTime);
+        removeLines(id, account, delegate, newAmount, blockTime);
+        return addLines(account, newDelegate, newAmount, newSlope, newCliff, blockTime);
     }
 
-    function totalSupply() public returns (uint) {
-        if (totalSupplyLine.initial.start == 0) { //no lock
+    function withdraw() external {
+        uint blockTime = roundTimestamp(block.timestamp);
```
Comment (evgenynacu) [inline staking/contracts/Staking.sol:117]: is it better to move this logic inside if (!stopped)?

File: staking/contracts/Staking.sol [hunk 2]
```diff
@@ -4,87 +4,250 @@ pragma solidity >=0.6.2 <0.8.0;
...
+    /**
+     * @dev counter for locks identifiers
+     */
+    uint public counter;
+
+    /**
+     * @dev true if contract entered not working state
+     */
+    bool private stopped;
+    address public migrateTo;               //address migrate to
```
Comment (evgenynacu) [inline staking/contracts/Staking.sol:38]: I think, we need to disable stake and other functions when migration started

File: staking/contracts/Staking.sol [hunk 3]
```diff
@@ -4,87 +4,250 @@ pragma solidity >=0.6.2 <0.8.0;
...
+        }
+        if (value > 0) {
+            locks[msg.sender].amount = locks[msg.sender].amount.sub(value);
+            require(token.transfer(msg.sender, value), "Failure while transferring, withdraw");
+        }
+    }
+
+    function delegate(uint id, address newDelegate) external notStopped {
+        address from = deposits[id].delegate;
+        require(from != address(0), "Delegate from address by id not found");
```
Comment (evgenynacu) [inline staking/contracts/Staking.sol:132]: this message doesn't tell anything to the user. why delegate can't be zero? may be you are trying to check if deposit exists?

File: staking/contracts/Staking.sol [hunk 4]
```diff
@@ -4,87 +4,250 @@ pragma solidity >=0.6.2 <0.8.0;
...
+        LibBrokenLine.LineData memory lineData = locks[from].balance.initiatedLines[id];
+        require(lineData.line.bias != 0, "Line already finished nothing to delegate");
```
Comment (evgenynacu) [inline staking/contracts/Staking.sol:134]: message is not good. user doesn't know anything about Lines

**Review by evgenynacu — CHANGES_REQUESTED** [review]

File: staking/test/contracts/StakingTest.sol [hunk 5]
```diff
@@ -9,23 +9,34 @@ contract StakingTest {
     event createLockResult(uint result);
     event balanceOfResult(uint result);
     event totalBalanceResult(uint result);
+    event reStakeResult(uint result);
 
-    function _createLock(address staking, address account, uint amount, uint period, uint cliff) external {
+    function _stake(address staking, address locker, address delegate, uint amount, uint slope, uint cliff) external {
```
Comment (evgenynacu) [inline staking/test/contracts/StakingTest.sol:14]: Do you need this contract? Why Stake can't be used without this StakingTest?

File: staking/contracts/Staking.sol [hunk 6]
```diff
@@ -4,87 +4,250 @@ pragma solidity >=0.6.2 <0.8.0;
...
+        uint blockTime = roundTimestamp(block.timestamp);
+        (uint bias, uint slope) = locks[from].balance.remove(id, blockTime);
+        LibBrokenLine.Line memory line = LibBrokenLine.Line(blockTime, bias, slope);
+        uint cliff = lineData.cliff;
```
Comment (evgenynacu) [inline staking/contracts/Staking.sol:138]: is cliff initial cliff from the line?
How will this case work
1. create Line with cliff
2. wait for cliff to be pass
3. try to delegate

Will cliff be added in the start?
Is it a test like this? If not, I think, it needs to be added

**Review by evgenynacu — CHANGES_REQUESTED** [review]

File: staking/contracts/Staking.sol [hunk 7]
```diff
@@ -4,87 +4,250 @@ pragma solidity >=0.6.2 <0.8.0;
...
+        if (addAmount > balance) {//need more, than balance, so need transfer ERC20 to this
+            require(token.transferFrom(deposits[id].locker, address(this), addAmount.sub(balance)), "Failure while transferring");
+            locks[account].amount = locks[account].amount.sub(residue);
+            locks[account].amount = locks[account].amount.add(newAmount);
+        }
+        locks[delegate].balance.remove(id, toTime);
+        totalSupplyLine.remove(id, toTime);
+    }
 
+    function addLines(address account, address newDelegate, uint newAmount, uint newSlope, uint newCliff, uint blockTime) internal returns (uint) {
```
Comment (evgenynacu) [inline staking/contracts/Staking.sol:217]: this function looks like stake. Is it better to invoke addLines from stake?

File: staking/test/contracts/StakingTest.sol (hunk 5 above)
Comment (SKonstantinS) [reply staking/test/contracts/StakingTest.sol:14]: I use event emit to return resultsof contract Staking methods  in *test.js. I cant found another way. So I need StakingTest.

**Review by evgenynacu — CHANGES_REQUESTED** [review]
more changes

File: staking/contracts/Staking.sol (hunk 1 above)
Comment (SKonstantinS) [reply staking/contracts/Staking.sol:117]: yes better to move logic in if()
agree. do it

File: staking/contracts/Staking.sol (hunk 2 above)
Comment (SKonstantinS) [reply staking/contracts/Staking.sol:38]: Lets think:
migration  - long process and imagine: migration start, and for example ordinary life go on one or two weeks. 
If everything is all right with new contract(migrate to), and over 50% users already migrate so only after that set status stopped. Anyway we(contract owner) can call stop() before startMigration() we have this mechanism now, but
I think, we should not collect migration and stop in hard code.
My opinion, If @evgenynacu need, I do

File: staking/contracts/Staking.sol (hunk 3 above)
Comment (SKonstantinS) [reply staking/contracts/Staking.sol:132]: better:  deposit not exists
agree, do

File: staking/contracts/Staking.sol (hunk 4 above)
Comment (SKonstantinS) [reply staking/contracts/Staking.sol:134]: better: deposit already finished
agree? do

File: staking/contracts/Staking.sol (hunk 6 above)
Comment (SKonstantinS) [reply staking/contracts/Staking.sol:138]: cliff initial cliff from the line? - yes

Error !!!
I have test about  it, Test2, but cliff=0 and test don`t show error
Write test with cliff = 2, cliff already finished, but method set it from initial value

File: staking/contracts/Staking.sol (hunk 7 above)
Comment (SKonstantinS) [reply staking/contracts/Staking.sol:217]: agree, do it

**Comment by SKonstantinS** [comment]
not actual now, close

**Comment by SKonstantinS** [comment]
not actual close#2

**Comment by SKonstantinS** [comment]
third attempt to close

**Pull request closed without merging**